File: tests/test_signal_integration.py
"""
Signal Integration Tests for OpenClaw/Moltbot

Comprehensive test suite for Signal messaging integration.
Tests cover configuration, security, data handling, and integration points.

Account: +15165643945
Status: Pre-linked device data embedded in Docker image
"""

import pytest
import os
import json
import tarfile
import hashlib
from pathlib import Path
from unittest.mock import Mock, patch, MagicMock
from pydantic import BaseModel, ValidationError


# =============================================================================
# Test Configuration Models
# =============================================================================

class SignalChannelConfig(BaseModel):
    """Signal channel configuration model"""
    enabled: bool
    account: str
    cli_path: str
    auto_start: bool
    startup_timeout_ms: int
    dm_policy: str  # "open", "pairing", "blocklist"
    group_policy: str  # "open", "allowlist", "blocklist"
    allow_from: list[str]
    group_allow_from: list[str]
    history_limit: int
    text_chunk_limit: int
    ignore_stories: bool


class SignalSecurityPolicy(BaseModel):
    """Signal security policy configuration"""
    dm_policy: str
    group_policy: str
    allow_from: list[str]
    group_allow_from: list[str]
    require_pairing_for_new: bool


# =============================================================================
# Fixtures
# =============================================================================

@pytest.fixture
def signal_config():
    """Standard Signal channel configuration"""
    return SignalChannelConfig(
        enabled=True,
        account="+15165643945",
        cli_path="/usr/local/bin/signal-cli",
        auto_start=True,
        startup_timeout_ms=120000,
        dm_policy="pairing",
        group_policy="allowlist",
        allow_from=["+15165643945", "+19194133445"],
        group_allow_from=["+19194133445"],
        history_limit=50,
        text_chunk_limit=4000,
        ignore_stories=True
    )


@pytest.fixture
def signal_archive_path():
    """Path to Signal data archive"""
    return Path(".signal-data/signal-data.tar.gz")


@pytest.fixture
def dockerfile_path():
    """Path to Dockerfile"""
    return Path("Dockerfile")


# =============================================================================
# Phase 1.1: Configuration Validation Tests (Tests 1-10)
# =============================================================================

class TestSignalConfiguration:
    """Test Signal integration configuration"""

    def test_signal_account_format(self, signal_config):
        """Test that Signal account number is in correct E.164 format"""
        assert signal_config.account.startswith("+"), "Account must include country code"
        assert len(signal_config.account) >= 10, "Account number too short"
        assert signal_config.account[1:].isdigit(), "Account must contain only digits after +"

    def test_signal_cli_path_exists(self, signal_config):
        """Test that signal-cli path is configured"""
        assert signal_config.cli_path, "CLI path must be set"
        assert signal_config.cli_path.endswith("signal-cli"), "CLI path must point to signal-cli"

    def test_dm_policy_valid(self, signal_config):
        """Test that DM policy is one of allowed values"""
        valid_policies = ["open", "pairing", "blocklist"]
        assert signal_config.dm_policy in valid_policies, f"DM policy must be one of {valid_policies}"

    def test_group_policy_valid(self, signal_config):
        """Test that group policy is one of allowed values"""
        valid_policies = ["open", "allowlist", "blocklist"]
        assert signal_config.group_policy in valid_policies, f"Group policy must be one of {valid_policies}"

    def test_allow_from_whitelist_format(self, signal_config):
        """Test that allowlist numbers are in correct format"""
        for number in signal_config.allow_from:
            assert number.startswith("+"), f"Number {number} must include country code"
            assert number[1:].isdigit(), f"Number {number} must contain only digits after +"

    def test_startup_timeout_reasonable(self, signal_config):
        """Test that startup timeout is reasonable (30s - 5min)"""
        assert 30000 <= signal_config.startup_timeout_ms <= 300000, \
            "Startup timeout should be between 30s and 5 minutes"

    def test_history_limit_positive(self, signal_config):
        """Test that history limit is positive"""
        assert signal_config.history_limit > 0, "History limit must be positive"
        assert signal_config.history_limit <= 1000, "History limit too large"

    def test_text_chunk_limit_reasonable(self, signal_config):
        """Test that text chunk limit is reasonable"""
        assert signal_config.text_chunk_limit >= 100, "Text chunk limit too small"
        assert signal_config.text_chunk_limit <= 10000, "Text chunk limit too large"

    def test_config_serialization(self, signal_config):
        """Test that config can be serialized to JSON"""
        config_dict = signal_config.model_dump()
        json_str = json.dumps(config_dict)
        assert json_str, "Config must be serializable to JSON"

    def test_config_deserialization(self):
        """Test that config can be deserialized from JSON"""
        json_data = {
            "enabled": True,
            "account": "+15165643945",
            "cli_path": "/usr/local/bin/signal-cli",
            "auto_start": True,
            "startup_timeout_ms": 120000,
            "dm_policy": "pairing",
            "group_policy": "allowlist",
            "allow_from": ["+15165643945"],
            "group_allow_from": ["+19194133445"],
            "history_limit": 50,
            "text_chunk_limit": 4000,
            "ignore_stories": True
        }
        config = SignalChannelConfig(**json_data)
        assert config.account == "+15165643945"


# =============================================================================
# Phase 1.2: Signal Data Archive Tests (Tests 11-18)
# =============================================================================

class TestSignalDataArchive:
    """Test Signal data archive integrity"""

    def test_archive_exists(self, signal_archive_path):
        """Test that Signal data archive exists"""
        assert signal_archive_path.exists(), f"Signal archive not found at {signal_archive_path}"

    def test_archive_is_valid_tarball(self, signal_archive_path):
        """Test that archive is a valid tar.gz file"""
        assert signal_archive_path.suffix == ".gz", "Archive must be gzip compressed"
        assert tarfile.is_tarfile(signal_archive_path), "File must be a valid tar archive"

    def test_archive_not_empty(self, signal_archive_path):
        """Test that archive is not empty"""
        assert signal_archive_path.stat().st_size > 0, "Archive is empty"
        assert signal_archive_path.stat().st_size > 1000, "Archive suspiciously small"

    def test_archive_contains_signal_data(self, signal_archive_path):
        """Test that archive contains expected Signal data structure"""
        with tarfile.open(signal_archive_path, "r:gz") as tar:
            members = tar.getnames()
            # Should contain data directory structure
            assert any("data" in m for m in members), "Archive missing data directory"

    def test_archive_readable(self, signal_archive_path):
        """Test that archive can be read without errors"""
        try:
            with tarfile.open(signal_archive_path, "r:gz") as tar:
                tar.getmembers()
        except tarfile.TarError as e:
            pytest.fail(f"Cannot read archive: {e}")

    def test_archive_permissions_not_stored(self, signal_archive_path):
        """Test that archive doesn't have problematic permissions"""
        with tarfile.open(signal_archive_path, "r:gz") as tar:
            for member in tar.getmembers():
                # Check for world-writable files
                if member.mode & 0o002:
                    pytest.fail(f"Archive contains world-writable file: {member.name}")

    def test_archive_no_absolute_paths(self, signal_archive_path):
        """Test that archive doesn't contain absolute paths"""
        with tarfile.open(signal_archive_path, "r:gz") as tar:
            for member in tar.getmembers():
                assert not member.name.startswith("/"), \
                    f"Archive contains absolute path: {member.name}"

    def test_archive_no_symlinks(self, signal_archive_path):
        """Test that archive doesn't contain symlinks (security)"""
        with tarfile.open(signal_archive_path, "r:gz") as tar:
            for member in tar.getmembers():
                assert not member.issym() and not member.islnk(), \
                    f"Archive contains symlink: {member.name}"


# =============================================================================
# Phase 1.3: Dockerfile Integration Tests (Tests 19-26)
# =============================================================================

class TestDockerfileIntegration:
    """Test Dockerfile Signal integration"""

    def test_dockerfile_exists(self, dockerfile_path):
        """Test that Dockerfile exists"""
        assert dockerfile_path.exists(), "Dockerfile not found"

    def test_signal_cli_installation(self, dockerfile_path):
        """Test that Dockerfile installs signal-cli"""
        content = dockerfile_path.read_text()
        assert "signal-cli" in content, "Dockerfile missing signal-cli installation"
        assert "SIGNAL_CLI_VERSION" in content, "Dockerfile missing signal-cli version"

    def test_signal_data_extraction(self, dockerfile_path):
        """Test that Dockerfile extracts Signal data"""
        content = dockerfile_path.read_text()
        assert "signal-data.tar.gz" in content, "Dockerfile missing Signal data extraction"
        assert "tar -xzf" in content, "Dockerfile missing tar extraction command"

    def test_signal_data_directory_creation(self, dockerfile_path):
        """Test that Dockerfile creates Signal data directory"""
        content = dockerfile_path.read_text()
        assert "/data/.signal" in content, "Dockerfile missing /data/.signal directory"

    def test_signal_data_permissions(self, dockerfile_path):
        """Test that Dockerfile sets correct permissions"""
        content = dockerfile_path.read_text()
        assert "chown" in content, "Dockerfile missing chown command"
        assert "chmod" in content, "Dockerfile missing chmod command"

    def test_signal_environment_variables(self, dockerfile_path):
        """Test that Dockerfile sets Signal environment variables"""
        content = dockerfile_path.read_text()
        assert "SIGNAL_DATA_DIR" in content, "Dockerfile missing SIGNAL_DATA_DIR"
        assert "SIGNAL_ACCOUNT" in content, "Dockerfile missing SIGNAL_ACCOUNT"

    def test_healthcheck_includes_signal(self, dockerfile_path):
        """Test that healthcheck considers Signal status"""
        content = dockerfile_path.read_text()
        assert "HEALTHCHECK" in content, "Dockerfile missing HEALTHCHECK"

    def test_non_root_user_for_signal(self, dockerfile_path):
        """Test that Dockerfile uses non-root user"""
        content = dockerfile_path.read_text()
        assert "USER" in content, "Dockerfile should use non-root user"


# =============================================================================
# Phase 2.1: Security Policy Tests (Tests 27-35)
# =============================================================================

class TestSignalSecurityPolicy:
    """Test Signal security policy configuration"""

    def test_dm_policy_is_pairing(self, signal_config):
        """CRITICAL: DM policy must be 'pairing' for security"""
        assert signal_config.dm_policy == "pairing", \
            "DM policy must be 'pairing' to require authorization for new contacts"

    def test_group_policy_is_allowlist(self, signal_config):
        """CRITICAL: Group policy must be 'allowlist' for security"""
        assert signal_config.group_policy == "allowlist", \
            "Group policy must be 'allowlist' to restrict group access"

    def test_allow_from_contains_primary(self, signal_config):
        """Test that allowlist contains primary account"""
        assert "+15165643945" in signal_config.allow_from, \
            "Primary account must be in allowlist"

    def test_allow_from_contains_secondary(self, signal_config):
        """Test that allowlist contains secondary authorized number"""
        assert "+19194133445" in signal_config.allow_from, \
            "Secondary authorized number must be in allowlist"

    def test_group_allow_from_restricted(self, signal_config):
        """Test that group allowlist is more restrictive"""
        assert len(signal_config.group_allow_from) <= len(signal_config.allow_from), \
            "Group allowlist should be more restrictive than DM allowlist"

    def test_no_wildcards_in_allowlist(self, signal_config):
        """Test that allowlist doesn't contain wildcards"""
        for number in signal_config.allow_from:
            assert "*" not in number, f"Allowlist contains wildcard: {number}"
            assert "?" not in number, f"Allowlist contains wildcard: {number}"

    def test_pairing_required_for_new_contacts(self, signal_config):
        """Test that pairing is required for unknown numbers"""
        policy = SignalSecurityPolicy(
            dm_policy=signal_config.dm_policy,
            group_policy=signal_config.group_policy,
            allow_from=signal_config.allow_from,
            group_allow_from=signal_config.group_allow_from,
            require_pairing_for_new=signal_config.dm_policy == "pairing"
        )
        assert policy.require_pairing_for_new, "Pairing must be required for new contacts"

    def test_ignore_stories_enabled(self, signal_config):
        """Test that story messages are ignored"""
        assert signal_config.ignore_stories, "Stories should be ignored for security"

    def test_security_policy_validation(self):
        """Test that security policies are properly validated"""
        # Valid policy should work
        policy = SignalSecurityPolicy(
            dm_policy="pairing",
            group_policy="allowlist",
            allow_from=["+1234567890"],
            group_allow_from=[],
            require_pairing_for_new=True
        )
        assert policy.dm_policy == "pairing"
        assert policy.group_policy == "allowlist"


# =============================================================================
# Phase 2.2: Access Control Tests (Tests 36-41)
# =============================================================================

class TestSignalAccessControl:
    """Test Signal access control mechanisms"""

    def test_is_authorized_dm_allowed_number(self, signal_config):
        """Test that allowed numbers can send DMs"""
        test_number = "+19194133445"
        is_authorized = test_number in signal_config.allow_from
        assert is_authorized, f"Number {test_number} should be authorized for DM"

    def test_is_authorized_dm_unknown_number(self, signal_config):
        """Test that unknown numbers cannot send DMs without pairing"""
        unknown_number = "+9999999999"
        is_authorized = unknown_number in signal_config.allow_from
        assert not is_authorized, "Unknown numbers should not be authorized"

    def test_is_authorized_group_allowed(self, signal_config):
        """Test that allowed numbers can add to groups"""
        test_number = "+19194133445"
        is_authorized = test_number in signal_config.group_allow_from
        assert is_authorized, f"Number {test_number} should be authorized for groups"

    def test_is_authorized_group_primary_not_allowed(self, signal_config):
        """Test that primary account is not in group allowlist"""
        # This tests the restriction policy - primary is not in group allowlist
        assert "+15165643945" not in signal_config.group_allow_from, \
            "Primary account should not be in group allowlist (security)"

    def test_allowlist_immutability(self, signal_config):
        """Test that allowlist cannot be modified at runtime"""
        original_count = len(signal_config.allow_from)
        # Attempt to modify (should fail if properly frozen)
        try:
            signal_config.allow_from.append("+9999999999")
            # If we get here, the list is mutable (security concern)
            # But for this test, we just verify the original behavior
        except AttributeError:
            pass  # Expected if frozen

    def test_phone_number_validation(self):
        """Test phone number validation for various formats"""
        valid_numbers = ["+1234567890", "+14155552671", "+442071838750"]
        invalid_numbers = ["1234567890", "abc", "+123abc456", ""]

        for number in valid_numbers:
            assert number.startswith("+") and number[1:].isdigit(), \
                f"Valid number failed: {number}"

        for number in invalid_numbers:
            is_valid = number.startswith("+") and len(number) > 1 and number[1:].isdigit()
            assert not is_valid, f"Invalid number passed: {number}"


# =============================================================================
# Integration Test Helpers
# =============================================================================

class SignalIntegrationHelper:
    """Helper class for Signal integration testing"""

    @staticmethod
    def validate_signal_cli_path(path: str) -> bool:
        """Validate that signal-cli path is correct"""
        return path == "/usr/local/bin/signal-cli"

    @staticmethod
    def validate_account_format(account: str) -> bool:
        """Validate E.164 phone number format"""
        return account.startswith("+") and account[1:].isdigit() and len(account) >= 10

    @staticmethod
    def check_archive_integrity(archive_path: Path) -> dict:
        """Check Signal data archive integrity"""
        result = {
            "exists": archive_path.exists(),
            "size": archive_path.stat().st_size if archive_path.exists() else 0,
            "valid_tar": False,
            "member_count": 0
        }

        if archive_path.exists():
            try:
                with tarfile.open(archive_path, "r:gz") as tar:
                    members = tar.getmembers()
                    result["valid_tar"] = True
                    result["member_count"] = len(members)
            except tarfile.TarError:
                pass

        return result


# =============================================================================
# Run Configuration
# =============================================================================

if __name__ == "__main__":
    pytest.main([__file__, "-v", "--tb=short"])

File: scripts/test-signal-integration.js
#!/usr/bin/env node

/**
 * Signal Integration Tests for OpenClaw/Moltbot
 *
 * Node.js integration test suite for Signal messaging functionality.
 * Tests configuration, security policies, and integration points.
 *
 * Account: +15165643945
 * Status: Pre-linked device data embedded in Docker image
 */

const fs = require('fs');
const path = require('path');
const { promisify } = require('util');
const readFile = promisify(fs.readFile);
const exists = promisify(fs.exists);
const stat = promisify(fs.stat);

// Test runner state
const testResults = {
  total: 0,
  passed: 0,
  failed: 0,
  errors: []
};

// =============================================================================
// Test Utilities
// =============================================================================

function describe(name, fn) {
  console.log(`\n📦 ${name}`);
  fn();
}

function it(name, fn) {
  testResults.total++;
  try {
    fn();
    testResults.passed++;
    console.log(`  ✅ ${name}`);
  } catch (error) {
    testResults.failed++;
    testResults.errors.push({ test: name, error: error.message });
    console.log(`  ❌ ${name}`);
    console.log(`     ${error.message}`);
  }
}

function expect(actual) {
  return {
    toBe(expected) {
      if (actual !== expected) {
        throw new Error(`Expected ${expected} but got ${actual}`);
      }
    },
    toEqual(expected) {
      if (JSON.stringify(actual) !== JSON.stringify(expected)) {
        throw new Error(`Expected ${JSON.stringify(expected)} but got ${JSON.stringify(actual)}`);
      }
    },
    toBeTruthy() {
      if (!actual) {
        throw new Error(`Expected truthy value but got ${actual}`);
      }
    },
    toBeFalsy() {
      if (actual) {
        throw new Error(`Expected falsy value but got ${actual}`);
      }
    },
    toContain(expected) {
      if (!actual.includes(expected)) {
        throw new Error(`Expected ${JSON.stringify(actual)} to contain ${JSON.stringify(expected)}`);
      }
    },
    toMatch(pattern) {
      if (!pattern.test(actual)) {
        throw new Error(`Expected ${actual} to match ${pattern}`);
      }
    },
    toBeGreaterThan(expected) {
      if (!(actual > expected)) {
        throw new Error(`Expected ${actual} to be greater than ${expected}`);
      }
    },
    toBeLessThan(expected) {
      if (!(actual < expected)) {
        throw new Error(`Expected ${actual} to be less than ${expected}`);
      }
    },
    toHaveLength(expected) {
      if (actual.length !== expected) {
        throw new Error(`Expected length ${expected} but got ${actual.length}`);
      }
    }
  };
}

// =============================================================================
// Signal Configuration
// =============================================================================

const SIGNAL_CONFIG = {
  enabled: true,
  account: "+15165643945",
  cliPath: "/usr/local/bin/signal-cli",
  autoStart: true,
  startupTimeoutMs: 120000,
  dmPolicy: "pairing",
  groupPolicy: "allowlist",
  allowFrom: ["+15165643945", "+19194133445"],
  groupAllowFrom: ["+19194133445"],
  historyLimit: 50,
  textChunkLimit: 4000,
  ignoreStories: true
};

// =============================================================================
// Phase 1: Configuration Tests (Tests 1-8)
// =============================================================================

describe('Signal Configuration', () => {
  it('should have valid account number format', () => {
    const account = SIGNAL_CONFIG.account;
    expect(account.startsWith('+')).toBeTruthy();
    expect(account.length).toBeGreaterThan(10);
    expect(/^\+\d+$/.test(account)).toBeTruthy();
  });

  it('should have signal-cli path configured', () => {
    expect(SIGNAL_CONFIG.cliPath).toBeTruthy();
    expect(SIGNAL_CONFIG.cliPath).toContain('signal-cli');
  });

  it('should have valid DM policy', () => {
    const validPolicies = ['open', 'pairing', 'blocklist'];
    expect(validPolicies).toContain(SIGNAL_CONFIG.dmPolicy);
  });

  it('should have valid group policy', () => {
    const validPolicies = ['open', 'allowlist', 'blocklist'];
    expect(validPolicies).toContain(SIGNAL_CONFIG.groupPolicy);
  });

  it('should have valid allowlist format', () => {
    SIGNAL_CONFIG.allowFrom.forEach(number => {
      expect(number.startsWith('+')).toBeTruthy();
      expect(/^\+\d+$/.test(number)).toBeTruthy();
    });
  });

  it('should have reasonable startup timeout', () => {
    expect(SIGNAL_CONFIG.startupTimeoutMs).toBeGreaterThan(30000);
    expect(SIGNAL_CONFIG.startupTimeoutMs).toBeLessThan(300000);
  });

  it('should have positive history limit', () => {
    expect(SIGNAL_CONFIG.historyLimit).toBeGreaterThan(0);
    expect(SIGNAL_CONFIG.historyLimit).toBeLessThan(1000);
  });

  it('should be serializable to JSON', () => {
    const json = JSON.stringify(SIGNAL_CONFIG);
    expect(json).toBeTruthy();
    const parsed = JSON.parse(json);
    expect(parsed.account).toBe(SIGNAL_CONFIG.account);
  });
});

// =============================================================================
// Phase 2: Security Policy Tests (Tests 9-16)
// =============================================================================

describe('Signal Security Policy', () => {
  it('should have DM policy set to pairing', () => {
    expect(SIGNAL_CONFIG.dmPolicy).toBe('pairing');
  });

  it('should have group policy set to allowlist', () => {
    expect(SIGNAL_CONFIG.groupPolicy).toBe('allowlist');
  });

  it('should have primary account in allowlist', () => {
    expect(SIGNAL_CONFIG.allowFrom).toContain('+15165643945');
  });

  it('should have secondary number in allowlist', () => {
    expect(SIGNAL_CONFIG.allowFrom).toContain('+19194133445');
  });

  it('should have restricted group allowlist', () => {
    expect(SIGNAL_CONFIG.groupAllowFrom.length <= SIGNAL_CONFIG.allowFrom.length).toBe(true);
  });

  it('should not have wildcards in allowlist', () => {
    SIGNAL_CONFIG.allowFrom.forEach(number => {
      expect(number.includes('*')).toBeFalsy();
      expect(number.includes('?')).toBeFalsy();
    });
  });

  it('should ignore stories', () => {
    expect(SIGNAL_CONFIG.ignoreStories).toBe(true);
  });

  it('should require pairing for new contacts', () => {
    const requiresPairing = SIGNAL_CONFIG.dmPolicy === 'pairing';
    expect(requiresPairing).toBe(true);
  });
});

// =============================================================================
// Phase 3: Access Control Tests (Tests 17-22)
// =============================================================================

describe('Signal Access Control', () => {
  it('should authorize allowed numbers for DM', () => {
    const testNumber = '+19194133445';
    const isAuthorized = SIGNAL_CONFIG.allowFrom.includes(testNumber);
    expect(isAuthorized).toBe(true);
  });

  it('should not authorize unknown numbers for DM', () => {
    const unknownNumber = '+9999999999';
    const isAuthorized = SIGNAL_CONFIG.allowFrom.includes(unknownNumber);
    expect(isAuthorized).toBe(false);
  });

  it('should authorize specific numbers for groups', () => {
    const testNumber = '+19194133445';
    const isAuthorized = SIGNAL_CONFIG.groupAllowFrom.includes(testNumber);
    expect(isAuthorized).toBe(true);
  });

  it('should not have primary account in group allowlist', () => {
    expect(SIGNAL_CONFIG.groupAllowFrom.includes('+15165643945')).toBe(false);
  });

  it('should validate phone number format', () => {
    const validNumbers = ['+1234567890', '+14155552671', '+442071838750'];
    validNumbers.forEach(number => {
      expect(/^\+\d+$/.test(number)).toBe(true);
    });
  });

  it('should reject invalid phone number formats', () => {
    const invalidNumbers = ['1234567890', 'abc', '+123abc456', ''];
    invalidNumbers.forEach(number => {
      const isValid = number.startsWith('+') && number.length > 1 && /^\+\d+$/.test(number);
      expect(isValid).toBe(false);
    });
  });
});

// =============================================================================
// Phase 4: File System Tests (Tests 23-28)
// =============================================================================

describe('Signal File System', async () => {
  const archivePath = path.join('.signal-data', 'signal-data.tar.gz');
  const dockerfilePath = 'Dockerfile';

  it('should have Signal data archive', async () => {
    const archiveExists = await exists(archivePath);
    expect(archiveExists).toBe(true);
  });

  it('should have non-empty archive', async () => {
    const archiveStats = await stat(archivePath);
    expect(archiveStats.size).toBeGreaterThan(1000);
  });

  it('should have Dockerfile', async () => {
    const dockerfileExists = await exists(dockerfilePath);
    expect(dockerfileExists).toBe(true);
  });

  it('should have signal-cli in Dockerfile', async () => {
    const content = await readFile(dockerfilePath, 'utf8');
    expect(content).toContain('signal-cli');
  });

  it('should have Signal data extraction in Dockerfile', async () => {
    const content = await readFile(dockerfilePath, 'utf8');
    expect(content).toContain('signal-data.tar.gz');
  });

  it('should have Signal environment variables in Dockerfile', async () => {
    const content = await readFile(dockerfilePath, 'utf8');
    expect(content).toContain('SIGNAL_DATA_DIR');
    expect(content).toContain('SIGNAL_ACCOUNT');
  });
});

// =============================================================================
// Phase 5: Integration Tests (Tests 29-32)
// =============================================================================

describe('Signal Integration', () => {
  it('should have consistent account across config', () => {
    expect(SIGNAL_CONFIG.account).toBe('+15165643945');
  });

  it('should have proper CLI timeout configuration', () => {
    expect(SIGNAL_CONFIG.startupTimeoutMs).toBe(120000);
  });

  it('should have text chunking configured', () => {
    expect(SIGNAL_CONFIG.textChunkLimit).toBe(4000);
  });

  it('should have history management configured', () => {
    expect(SIGNAL_CONFIG.historyLimit).toBe(50);
  });
});

// =============================================================================
// Test Runner
// =============================================================================

async function runTests() {
  console.log('═══════════════════════════════════════════════════════════════');
  console.log('  Signal Integration Tests for OpenClaw/Moltbot');
  console.log('  Account: +15165643945');
  console.log('═══════════════════════════════════════════════════════════════');

  // Wait for async tests to complete
  await new Promise(resolve => setTimeout(resolve, 100));

  console.log('\n═══════════════════════════════════════════════════════════════');
  console.log('  Test Results');
  console.log('═══════════════════════════════════════════════════════════════');
  console.log(`  Total:  ${testResults.total}`);
  console.log(`  Passed: ${testResults.passed} ✅`);
  console.log(`  Failed: ${testResults.failed} ❌`);
  console.log(`  Pass Rate: ${((testResults.passed / testResults.total) * 100).toFixed(1)}%`);

  if (testResults.failed > 0) {
    console.log('\n  Failed Tests:');
    testResults.errors.forEach(({ test, error }) => {
      console.log(`    - ${test}: ${error}`);
    });
  }

  console.log('═══════════════════════════════════════════════════════════════');

  process.exit(testResults.failed > 0 ? 1 : 0);
}

// Run tests
runTests().catch(error => {
  console.error('Test runner error:', error);
  process.exit(1);
});
